application.go

```diff
@@ -57,3 +64,120 @@ func (ac *ApplicationClient) GetApplication(
 
 	return res.Application, nil
 }
+
+// GetApplicationsDelegatingToGateway returns the application addresses that are
+// delegating to the given gateway address.
+func (ac *ApplicationClient) GetApplicationsDelegatingToGateway(
+	ctx context.Context,
+	gatewayAddress string,
+	queryHeight uint64,
+) ([]string, error) {
+	allApplications, err := ac.GetAllApplications(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("GetApplicationsDelegatingToGateway: error getting all applications: %w", err)
+	}
+
+	gatewayDelegatingApplications := make([]string, 0)
+	for _, application := range allApplications {
+		appRing := ApplicationRing{Application: application}
+		// Get the gateways that are delegated to the application
+		// at the query height and check if the given gateway address is in the list.
+		gatewaysDelegatedTo := appRing.ringAddressesAtBlock(queryHeight)
+		if slices.Contains(gatewaysDelegatedTo, gatewayAddress) {
+			// The application is delegating to the given gateway address, add it to the list.
+			gatewayDelegatingApplications = append(gatewayDelegatingApplications, application.Address)
+		}
+	}
+
+	return gatewayDelegatingApplications, nil
+}
+
+type ApplicationRing struct {
+	types.Application
+	PublicKeyFetcher
+}
+
+// GetRing returns the ring for the application.
+// The ring is created using the application's public key and the public keys of
+// the gateways that are currently delegated from the application.
+func (a ApplicationRing) GetRing(
+	ctx context.Context,
+	queryHeight uint64,
+) (addressRing *ring.Ring, err error) {
+	// Get the gateway addresses that are delegated from the application
+	// at the query height.
+	currentGatewayAddresses := a.ringAddressesAtBlock(queryHeight)
+
+	ringAddresses := make([]string, 0)
+	ringAddresses = append(ringAddresses, a.Application.Address)
+
+	// If there are no current gateway addresses, use the application address as the ring address.
+	if len(currentGatewayAddresses) == 0 {
+		ringAddresses = append(ringAddresses, a.Application.Address)
+	} else {
+		ringAddresses = append(ringAddresses, currentGatewayAddresses...)
+	}
+
+	curve := ring_secp256k1.NewCurve()
+	ringPoints := make([]ringtypes.Point, 0, len(ringAddresses))
+
+	// Create a ring point for each address.
+	for _, address := range ringAddresses {
+		pubKey, err := a.PublicKeyFetcher.GetPubKeyFromAddress(ctx, address)
```

Comment:
Shouldn't we check whether `a.PublicKeyFetcher` is present? There seem to be no codepath initializing it (or ensuring that it has been initialized)

Reply:
Thank you, fixed.